File: src/main/java/com/tms/csp/ast/FVar.java
package com.tms.csp.ast;


import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;

public class FVar {

    FVars fVars;
    public Var vr;
    Exp xorParent;
    List<Exp> vvConstraints;
    List<Exp> complexConstraints;

    Boolean value;

    int score = -1;

    public FVar(Var vr) {
        this.vr = vr;
    }

    public void addConstraint(Exp constraint) {
        assert constraint.isComplex();

        if (!constraint.containsVar(vr)) {
            System.err.println("constraint: " + constraint);
            System.err.println("constraint: " + constraint.getClass());
            System.err.println("vr: " + vr);
            System.err.println();
        }
        assert constraint.containsVar(vr);
        if (constraint.isXor()) {
            assert xorParent == null;
            xorParent = constraint;
            assert xorParent.getVars().containsVar(vr);
        } else if (constraint.isVv()) {
            assert !constraint.isXor();
            addVV(constraint);
        } else if (constraint.isComplex()) {
            assert !constraint.isXor();
            addComplex(constraint);
        }
    }

    public void addVV(Exp constraint) {
        if (vvConstraints == null) {
            vvConstraints = new ArrayList<Exp>();
        }
        vvConstraints.add(constraint);
    }

    public int getVVCount() {
        if (vvConstraints == null) return 0;
        return vvConstraints.size();
    }

    public int getComplexCount() {
        if (complexConstraints == null) return 0;
        return complexConstraints.size();
    }

    public void addComplex(Exp constraint) {
        if (complexConstraints == null) {
            complexConstraints = new ArrayList<Exp>();
        }
        complexConstraints.add(constraint);
    }

    public int getScore() {
        if (score == -1) {
            int s = 0;

            if (xorParent != null) {
                int argCount = xorParent.asXor().getArgCount();
                s += 1000 * argCount;
            }
            s += getVVCount() * 100;

            s += getComplexCount() * 10;

            score = s;
        }
        return score;
    }

    public static void sort(List<FVar> fVars) {
        Collections.sort(fVars, F_VAR_COMPARATOR);
    }

    public static final Comparator<FVar> F_VAR_COMPARATOR = new Comparator<FVar>() {
        @Override
        public int compare(FVar v1, FVar v2) {
            Integer s1 = v1.getScore();
            Integer s2 = v2.getScore();
            return -s1.compareTo(s2);
        }
    };

    public void assign(boolean value) {
        assert this.value == null;
        this.value = value;
    }


    public FVars getFVars() {
        return fVars;
    }

    public void print() {
        System.err.println("vr[" + vr + "]  score[" + getScore() + "]");
    }

}

File: src/main/java/com/tms/csp/ast/FVar.kt
package com.tms.csp.ast


import java.util.*

class FVar(var vr: Var) {

    var fVars: FVars? = null
        internal set
    internal var xorParent: Exp? = null
    internal var vvConstraints: MutableList<Exp>? = null
    internal var complexConstraints: MutableList<Exp>? = null

    internal var value: Boolean? = null

    internal var score = -1

    val vvCount: Int
        get() = if (vvConstraints == null) 0 else vvConstraints!!.size


    val complexCount: Int
        get() = if (complexConstraints == null) 0 else complexConstraints!!.size

    fun addConstraint(constraint: Exp) {
        assert(constraint.isComplex)

        if (!constraint.containsVar(vr)) {
            System.err.println("constraint: $constraint")
            System.err.println("constraint: " + constraint.javaClass)
            System.err.println("vr: $vr")
            System.err.println()
        }
        assert(constraint.containsVar(vr))



        if (constraint.isXor) {
            assert(xorParent == null)
            xorParent = constraint
            assert(xorParent!!.vars.containsVar(vr))
            throw IllegalStateException()
        }
//        else if (constraint.isLitImp) {
//            assert(!constraint.isXor)
//            addVV(constraint)
//        }
        else if (constraint.isLitImp) {
            assert(!constraint.isXor)
            addVV(constraint)
        } else if (constraint.isComplex) {
            assert(!constraint.isXor)
            addComplex(constraint)
        }
    }

    fun addVV(constraint: Exp) {
        if (vvConstraints == null) {
            vvConstraints = ArrayList()
        }
        vvConstraints!!.add(constraint)
    }

    fun addComplex(constraint: Exp) {
        if (complexConstraints == null) {
            complexConstraints = ArrayList()
        }
        complexConstraints!!.add(constraint)
    }

    fun getScore(): Int {
        if (score == -1) {
            var s = 0

            if (xorParent != null) {
//                println("AAAAA")
//                val argCount = xorParent!!.asXor.argCount
//                s += 1000 * argCount
                throw IllegalStateException()
            }
            s += vvCount * 100

            s += complexCount * 10

            score = s
        }
        return score
    }

    fun assign(value: Boolean) {
        assert(this.value == null)
        this.value = value
    }

    fun print() {
        System.err.println("vr[" + vr + "]  score[" + getScore() + "]")
    }

    companion object {

        fun sort(fVars: List<FVar>) {
            Collections.sort(fVars, F_VAR_COMPARATOR)
        }

        val F_VAR_COMPARATOR: Comparator<FVar> = Comparator { v1, v2 ->
            val s1 = v1.getScore()
            val s2 = v2.getScore()
            -s1.compareTo(s2)
        }
    }

}

